File: code/javascript/hangman.mjs
import * as fs from "node:fs"
import * as readlinePromises from 'node:readline/promises';

import { HANGMAN_UI } from "./graphics.mjs";
import {ANSI} from "./ansi.mjs";

const rl = readlinePromises.createInterface({input: process.stdin,output: process.stdout});

const WORD_LIST = ["car", "cat", "dog", "elephant", "fish", "giraffe", "horse", "monkey", "panda", "sheep", "tiger", "zebra", "ant","bee","marmelade","computer","dax","sjøbanan"];

const logging = fs.createWriteStream("./hangman.log",{flags:"a"});

function logg(message){
    logging.write(message);
}

let atempts = 0;
const MAX_ATEMPTS = HANGMAN_UI.length-1;
const word = pickRandomItemFromList(WORD_LIST);
const wordLength = word.length;
const wordArray = word.split("");
const wordDisplay = createWordDisplay();

logg(word);

const guesses = [];
let isPlaing = true;

write(ANSI.CLEAR_SCREEN)

while(isPlaing){
    
    write(ANSI.BELL);
    write(ANSI.CLEAR_SCREEN)
    write(ANSI.CURSOR_HOME)
    write("Lets play Hangman!\n");
    write(`${ANSI.moveCursorTo(4,20)} Word: ${wordDisplay.join(" ")}\n`);
    write(`${ANSI.moveCursorTo(6,20)} Guesses: ${guesses.join(" ")}\n`);
    write(ANSI.CURSOR_HOME)
    write(`${HANGMAN_UI[atempts]}\n`);
    
    const guess = await rl.question("Your guess: ");
    guesses.push(guess);
    if(wordArray.includes(guess)){
        
        for(let index = 0; index < wordArray.length; index++){
            const letter = wordArray[index];
            if(letter === guess){
                wordDisplay[index] = `${ANSI.COLOR.GREEN}${letter}${ANSI.RESET}`;
            }
        }
       
    } else{
        atempts++;
    }

    if(atempts === MAX_ATEMPTS){
        isPlaing = false;
    }

}

write(`${ANSI.COLOR.RED}GAME OVER\n${ANSI.RESET}`)

setTimeout(() => {
    write(ANSI.CLEAR_SCREEN);
    write(ANSI.CURSOR_HOME);
    process.exit(0);
}, 3000);


function write(text) {
    // Noe alla console.log men uten "\n" på slutten av vær print.
    process.stdout.write(text);
}

function pickRandomItemFromList(list){
    return list[Math.floor(Math.random() * WORD_LIST.length)];
}

function createWordDisplay(wordLenght){
    return new Array(wordLength).fill("_")
}

File: code/python/hangman.py
from threading import Timer
import random, sys
import  graphics 
import ansi as  ANSI

HANGMAN_UI = graphics.HANGMAN_UI
WORD_LIST = ["car", "cat", "dog", "elephant", "fish", "giraffe", "horse", "monkey", "panda", "sheep", "tiger", "zebra", "ant","bee","marmelade","computer","dax","sjøbanan"]
MAX_ATEMPTS = len(HANGMAN_UI) #len er hvordan man finner lengden av variabler som lister og tekst med python
atempts = 0

word = random.choice(WORD_LIST)
wordLength = len(word)
wordArray = list(word) #Python har en split funksjon, men den splitter ikke til enkelt bokstaver, men vi kan konvertere direkte til en liste 
wordDisplay = ['_'] * wordLength

guesses = []
isPlaing = True


def write(text) :
    sys.stdout.write(text)

def exitGame() :
    write(ANSI.CLEAR_SCREEN)
    write(ANSI.CURSOR_HOME)
    sys.exit(0)


write(ANSI.CLEAR_SCREEN)

while isPlaing :

    write(ANSI.CLEAR_SCREEN)
    write(ANSI.moveCursorTo(0,0))
    write("Lets play Hangman!\n")
    write(f'{ANSI.moveCursorTo(4,20)} Word: {" ".join(wordDisplay)}\n')
    write(f'{ANSI.moveCursorTo(6,20)} Guesses: {" ".join(guesses)}\n')
    write(ANSI.CURSOR_HOME)
    write(f'{HANGMAN_UI[atempts]}\n')

    write("Your guess: ")
    guess = input()
    guesses.append(guess)

    if guess in wordArray :
        # Denne er ikke like frem å forstå, men vi trennger en index for å kunne sette verdier 
        # i wordDisplay listen, derfor lager vi to varabler index (posisjon i listen som git av enumerate)
        # og letter som er det faktiske tegnet vi skal sjekke mot.
        for index, letter in  enumerate(wordArray) :
            if letter == guess :
                wordDisplay[index] = f'{ANSI.GREEN}{letter}{ANSI.RESET}'
    else :
        atempts += 1

    if atempts == MAX_ATEMPTS :
        isPlaing = False

write(f'{ANSI.RED}GAME OVER\n{ANSI.RESET}')



timer = Timer(3.0, exitGame)
timer.start()
